# XGBoost di paling bawah

In [130]:
import pandas as pd
import numpy as np
url = 'https://raw.githubusercontent.com/farrelrassya/teachingMLDL/refs/heads/main/02.%20Deep%20Learning/Dataset/Infrared.csv'
df = pd.read_csv(url)
df.head()


Gender    Age                  Ethnicity  T_atm  Humidity  Distance  \
0    Male  41-50                      White   24.0      28.0       0.8   
1  Female  31-40  Black or African-American   24.0      26.0       0.8   
2  Female  21-30                      White   24.0      26.0       0.8   
3  Female  21-30  Black or African-American   24.0      27.0       0.8   
4    Male  18-20                      White   24.0      27.0       0.8   

   T_offset1  Max1R13_1  Max1L13_1  aveAllR13_1  ...  T_FHRC1  T_FHLC1  \
0     0.7025    35.0300    35.3775      34.4000  ...  33.4775  33.3725   
1     0.7800    34.5500    34.5200      33.9300  ...  34.0550  33.6775   
2     0.8625    35.6525    35.5175      34.2775  ...  34.8275  34.6475   
3     0.9300    35.2225    35.6125      34.3850  ...  34.4225  34.6550   
4     0.8950    35.5450    35.6650      34.9100  ...  35.1600  34.3975   

   T_FHBC1  T_FHTC1  T_FH_Max1  T_FHC_Max1   T_Max1    T_OR1  T_OR_Max1  \
0  33.4925  33.0025    34.5300     34.0075  35.6925  35.6350    35.6525   
1  33.9700  34.0025    34.6825     34.6600  35.1750  35.0925    35.1075   
2  34.8200  34.6700    35.3450     35.2225  35.9125  35.8600    35.8850   
3  34.3025  34.9175    35.6025     35.3150  35.7200  34.9650    34.9825   
4  34.6700  33.8275    35.4175     35.3725  35.8950  35.5875    35.6175   

   aveOralM  
0     36.59  
1     37.19  
2     37.34  
3     37.09  
4     37.04  

[5 rows x 34 columns]

In [131]:
df = df.dropna()

In [132]:
df.dtypes

Gender          object
Age             object
Ethnicity       object
T_atm          float64
Humidity       float64
Distance       float64
T_offset1      float64
Max1R13_1      float64
Max1L13_1      float64
aveAllR13_1    float64
aveAllL13_1    float64
T_RC1          float64
T_RC_Dry1      float64
T_RC_Wet1      float64
T_RC_Max1      float64
T_LC1          float64
T_LC_Dry1      float64
T_LC_Wet1      float64
T_LC_Max1      float64
RCC1           float64
LCC1           float64
canthiMax1     float64
canthi4Max1    float64
T_FHCC1        float64
T_FHRC1        float64
T_FHLC1        float64
T_FHBC1        float64
T_FHTC1        float64
T_FH_Max1      float64
T_FHC_Max1     float64
T_Max1         float64
T_OR1          float64
T_OR_Max1      float64
aveOralM       float64
dtype: object

In [133]:
df['Ethnicity'].value_counts()

Ethnicity
White                                504
Asian                                260
Black or African-American            143
Hispanic/Latino                       57
Multiracial                           50
American Indian or Alaskan Native      4
Name: count, dtype: int64

In [134]:
df['Age'].value_counts()

Age
18-20    532
21-25    355
26-30     67
31-40     31
51-60     11
21-30     10
41-50      9
>60        3
Name: count, dtype: int64

##### Encode the Object datatypes, for ordinal data use label encoder (or manual), and for non-ordinal data use one-hot encoder

In [135]:
gender_mapping = {
    'Male': 0,
    'Female': 1
}

df['Gender'] = df['Gender'].map(gender_mapping)
df['Gender'].value_counts()


Gender
1    605
0    413
Name: count, dtype: int64

Mapping labelling dengan manual agar jarak age sesuai dengan ordinalitasnya

In [136]:
age_mapping = {
    '18-20': 0,
    '21-25': 1, 
    '21-30': 2,
    '26-30': 3,
    '31-40': 4,
    '41-50': 5,
    '51-60': 6,
    '>60': 7
}

df['Age'] = df['Age'].map(age_mapping)
df['Age'].value_counts()




Age
0    532
1    355
3     67
4     31
6     11
2     10
5      9
7      3
Name: count, dtype: int64

Data Ethnicity disini menggunakan one-hot karena datanya tidak ordinal

In [137]:
ethnicity_encoded = pd.get_dummies(df['Ethnicity'], prefix='Ethnicity')
df = pd.concat([df, ethnicity_encoded], axis=1)
df = df.drop('Ethnicity', axis=1)
df.head()

Gender  Age  T_atm  Humidity  Distance  T_offset1  Max1R13_1  Max1L13_1  \
0       0    5   24.0      28.0       0.8     0.7025    35.0300    35.3775   
1       1    4   24.0      26.0       0.8     0.7800    34.5500    34.5200   
2       1    2   24.0      26.0       0.8     0.8625    35.6525    35.5175   
3       1    2   24.0      27.0       0.8     0.9300    35.2225    35.6125   
4       0    0   24.0      27.0       0.8     0.8950    35.5450    35.6650   

   aveAllR13_1  aveAllL13_1  ...   T_Max1    T_OR1  T_OR_Max1  aveOralM  \
0      34.4000      34.9175  ...  35.6925  35.6350    35.6525     36.59   
1      33.9300      34.2250  ...  35.1750  35.0925    35.1075     37.19   
2      34.2775      34.8000  ...  35.9125  35.8600    35.8850     37.34   
3      34.3850      35.2475  ...  35.7200  34.9650    34.9825     37.09   
4      34.9100      35.3675  ...  35.8950  35.5875    35.6175     37.04   

   Ethnicity_American Indian or Alaskan Native  Ethnicity_Asian  \
0                                        False            False   
1                                        False            False   
2                                        False            False   
3                                        False            False   
4                                        False            False   

   Ethnicity_Black or African-American  Ethnicity_Hispanic/Latino  \
0                                False                      False   
1                                 True                      False   
2                                False                      False   
3                                 True                      False   
4                                False                      False   

   Ethnicity_Multiracial  Ethnicity_White  
0                  False             True  
1                  False            False  
2                  False             True  
3                  False            False  
4                  False             True  

[5 rows x 39 columns]

In [138]:
from sklearn.model_selection import train_test_split
X = df.drop('aveOralM', axis=1)
y = df['aveOralM']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [139]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)


In [140]:
X_train

array([[ 8.21521223e-01,  1.59204238e-01,  2.55041451e-03, ...,
        -2.41904021e-01, -2.30304371e-01, -9.82946374e-01],
       [-1.21725401e+00, -6.66224104e-01,  4.47415574e-01, ...,
        -2.41904021e-01, -2.30304371e-01,  1.01734950e+00],
       [-1.21725401e+00,  1.59204238e-01,  2.55041451e-03, ...,
        -2.41904021e-01, -2.30304371e-01, -9.82946374e-01],
       ...,
       [ 8.21521223e-01,  1.59204238e-01,  7.66946077e-02, ...,
        -2.41904021e-01, -2.30304371e-01,  1.01734950e+00],
       [ 8.21521223e-01,  1.59204238e-01,  2.44930879e+00, ...,
        -2.41904021e-01, -2.30304371e-01, -9.82946374e-01],
       [-1.21725401e+00, -6.66224104e-01, -2.19882165e-01, ...,
        -2.41904021e-01,  4.34207998e+00, -9.82946374e-01]],
      shape=(814, 38))

#K-Nearest Neighbors (KNN) Regressor

KNN regression adalah metode non-parametrik yang memprediksi nilai target suatu data poin dengan mengambil rata-rata dari nilai target **k** tetangga terdekatnya dalam ruang fitur. Jarak antara titik biasanya dihitung menggunakan **Euclidean distance**:

$d(p,q) = \sqrt{\sum_{i=1}^{n} (p_i - q_i)^2}$

di mana pp dan qq adalah dua titik dalam ruang berdimensi n.

Untuk suatu titik x, prediksi dihitung sebagai:

$\hat{y}(x) = \frac{1}{k}\sum_{i=1}^{k} y_i$

di mana $y_i$ adalah nilai target dari k tetangga terdekat.

## Decision Tree Regressor

Decision tree regression membuat model berbentuk pohon keputusan berdasarkan fitur-fitur yang ada. Algoritma ini membagi data ke dalam subset berdasarkan ambang batas fitur yang meminimalkan varians nilai target dalam setiap subset.

Pada setiap node, kriteria pemisahan bertujuan untuk meminimalkan:

$MSE = \frac{1}{n} \sum_{i=1}^{n} (y_i - \bar{y})^2$

di mana $\bar{y}$ adalah rata-rata nilai target dalam node tersebut.

Prediksi akhir untuk leaf node adalah rata-rata nilai target dari data pelatihan dalam node tersebut:

$\hat{y} = \frac{1}{n_{leaf}} \sum_{i \in leaf} y_i$

## XGBoost Regressor

XGBoost (eXtreme Gradient Boosting) adalah algoritma boosting yang sangat efisien dan efektif untuk masalah regresi. XGBoost bekerja dengan prinsip ensemble learning, di mana model dibangun secara sekuensial dengan setiap model baru berusaha memperbaiki kesalahan dari model sebelumnya.

XGBoost menggunakan fungsi objektif yang terdiri dari loss function dan regularization:

$Obj = \sum_{i=1}^{n} l(y_i, \hat{y}_i) + \sum_{k=1}^{K} \Omega(f_k)$

di mana $l$ adalah fungsi loss, $\Omega$ adalah regularisasi, dan $f_k$ adalah pohon keputusan ke-k.

Keunggulan XGBoost meliputi penanganan data yang efisien, kemampuan regularisasi untuk mencegah overfitting, dan dukungan untuk komputasi paralel yang mempercepat proses pelatihan.


In [141]:
from sklearn.neighbors import KNeighborsRegressor
knn = KNeighborsRegressor(n_neighbors=5)
knn.fit(X_train, y_train)
y_pred_knn = knn.predict(X_test)

In [142]:
from sklearn.tree import DecisionTreeRegressor
dt = DecisionTreeRegressor(random_state=42)
dt.fit(X_train, y_train)
y_pred_dt = dt.predict(X_test)

In [143]:
from xgboost import XGBRegressor

xgb = XGBRegressor(
    random_state=42,
    n_estimators=100,
    learning_rate=0.1,
    max_depth=5
)

xgb.fit(X_train, y_train)

y_pred_xgb = xgb.predict(X_test)

## Evaluasi Model
 
Untuk mengevaluasi performa model regresi, kita menggunakan beberapa metrik:
 
### Mean Squared Error (MSE)
MSE mengukur rata-rata kuadrat selisih antara nilai prediksi dan nilai aktual:
 
$MSE = \frac{1}{n}\sum_{i=1}^{n}(y_i - \hat{y}_i)^2$
 
Semakin kecil nilai MSE, semakin baik model dalam memprediksi.
 
### Root Mean Squared Error (RMSE) 
RMSE adalah akar kuadrat dari MSE:

$RMSE = \sqrt{MSE} = \sqrt{\frac{1}{n}\sum_{i=1}^{n}(y_i - \hat{y}_i)^2}$

RMSE memiliki unit yang sama dengan variabel target, sehingga lebih mudah diinterpretasi.

### R-squared (R²)
R² mengukur proporsi variasi dalam variabel target yang dapat dijelaskan oleh model:
 
$R^2 = 1 - \frac{\sum_{i=1}^{n}(y_i - \hat{y}_i)^2}{\sum_{i=1}^{n}(y_i - \bar{y})^2}$
 
- R² bernilai antara 0 dan 1
- R² = 1 berarti model sempurna menjelaskan variasi data
- R² = 0 berarti model tidak lebih baik dari memprediksi rata-rata


In [144]:
from sklearn.metrics import mean_squared_error, r2_score
mse_knn = mean_squared_error(y_test, y_pred_knn)
rmse_knn = np.sqrt(mse_knn)
r2_knn = r2_score(y_test, y_pred_knn)

mse_dt = mean_squared_error(y_test, y_pred_dt)
rmse_dt = np.sqrt(mse_dt)
r2_dt = r2_score(y_test, y_pred_dt)

print(f"KNN: MSE={mse_knn:.4f}, RMSE={rmse_knn:.4f}, R2={r2_knn:.4f}")
print(f"Decision Tree: MSE={mse_dt:.4f}, RMSE={rmse_dt:.4f}, R2={r2_dt:.4f}")


KNN: MSE=0.0707, RMSE=0.2659, R2=0.6445
Decision Tree: MSE=0.1157, RMSE=0.3401, R2=0.4184


In [145]:
mse_xgb = mean_squared_error(y_test, y_pred_xgb)
rmse_xgb = np.sqrt(mse_xgb)
r2_xgb = r2_score(y_test, y_pred_xgb)

print(f"XGBoost: MSE={mse_xgb:.4f}, RMSE={rmse_xgb:.4f}, R2={r2_xgb:.4f}")

XGBoost: MSE=0.0618, RMSE=0.2486, R2=0.6893


## Kesimpulan

Berdasarkan hasil evaluasi model, dapat disimpulkan bahwa:

1. **XGBoost** menunjukkan performa terbaik dengan:
   - MSE terendah (0.0618)
   - RMSE terendah (0.2486)
   - R² tertinggi (0.6893), yang berarti model ini dapat menjelaskan sekitar 68.93% variasi dalam data target.

2. **KNN** menempati posisi kedua dengan:
   - MSE = 0.0707
   - RMSE = 0.2659
   - R² = 0.6445, menjelaskan sekitar 64.45% variasi data.

3. **Decision Tree** menunjukkan performa paling rendah dengan:
   - MSE tertinggi (0.1157)
   - RMSE tertinggi (0.3401)
   - R² terendah (0.4184), hanya menjelaskan sekitar 41.84% variasi data.

Secara keseluruhan, algoritma boosting (XGBoost) terbukti paling efektif untuk kasus regresi pada dataset infrared ini, mengungguli metode KNN dan Decision Tree.